Microsoft.Toolkit.Uwp.UI.Controls/UniformGrid/UniformGrid.Helpers.cs

```diff
@@ -0,0 +1,210 @@
+        #pragma warning disable SA1009 // Closing parenthesis must be followed by a space.
+        internal static IEnumerable<(int row, int column)> GetFreeSpot(TakenSpotsReferenceHolder arrayref, int firstcolumn, bool topdown)
+        #pragma warning restore SA1009 // Closing parenthesis must be followed by a space.
+        {
+            if (topdown)
+            {
+                var rows = arrayref.SpotsTaken.GetLength(0);
+
+                // Layout spots from Top-Bottom, Left-Right (right-left handled automatically by Grid with Flow-Direction).
+                // Effectively transpose the Grid Layout.
+                for (int c = 0; c < arrayref.SpotsTaken.GetLength(1); c++)
+                {
+                    int start = (c == 0 && firstcolumn > 0 && firstcolumn < rows) ? firstcolumn : 0;
+                    for (int r = start; r < rows; r++)
+                    {
+                        // TODO: Do we want/need to worry about size here, what is our expectation?
+                        if (!arrayref.SpotsTaken[r, c])
+                        {
+                            yield return (r, c);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                var columns = arrayref.SpotsTaken.GetLength(1);
+
+                // Layout spots as normal from Left-Right.
+                // (right-left handled automatically by Grid with Flow-Direction
+                // during its layout, internal model is always left-right).
+                for (int r = 0; r < arrayref.SpotsTaken.GetLength(0); r++)
+                {
+                    int start = (r == 0 && firstcolumn > 0 && firstcolumn < columns) ? firstcolumn : 0;
+                    for (int c = start; c < columns; c++)
+                    {
+                        // TODO: Do we want/need to worry about size here, what is our expectation?
+                        if (!arrayref.SpotsTaken[r, c])
+                        {
+                            yield return (r, c);
+                        }
+                    }
+                }
+            }
+        }
+
+        // Based on the number of visible children,
+        // returns the dimensions of the
+        // grid we need to hold all elements.
+        #pragma warning disable SA1008 // Opening parenthesis must be spaced correctly
+        internal static (int rows, int columns) GetDimensions(ref IEnumerable<FrameworkElement> visible, int rows, int cols, int firstColumn) //// TODO: Switch to 'in' for C# 7.2
+        #pragma warning restore SA1008 // Opening parenthesis must be spaced correctly
+        {
+            // If a dimension isn't specified, we need to figure out the other one (or both).
+            if (rows == 0 || cols == 0)
+            {
+                // Calculate the size & area of all objects in the grid to know how much space we need.
+                // TODO: Need to trim size of objects that go out of bounds?
```

Comment:
Can remove this comment, Grid does the 'trimming' automatically and called this out in the documentation.